File: ai_assist/providers/offline.py
"""Deterministic, network-free provider.

This is the default backend: it summarises a recognized diagram with simple
rules, so the assistant feature works out of the box with no API key.  It
also serves as the automatic fallback whenever a remote provider fails.
"""

from __future__ import annotations


class RuleProvider:
    name = "offline"

    def available(self) -> bool:
        return True

    def summarize(
        self,
        system_prompt: str,
        user_prompt: str,
        *,
        max_tokens: int = 1024,
    ) -> str:
        """Interface-compatible no-op; use :meth:`summarize_diagram` instead."""
        return ""

    @staticmethod
    def summarize_diagram(text_regions, shapes, latex: list[str]) -> str:
        """Build a human-readable summary from recognition output.

        ``text_regions`` items need a ``.text`` attribute, ``shapes`` items a
        ``.kind`` attribute.
        """
        if not shapes and not text_regions and not latex:
            return "The canvas is blank: no shapes, text or formulas were recognized."

        parts: list[str] = []
        if shapes:
            counts: dict[str, int] = {}
            for shape in shapes:
                kind = shape.kind or "unknown"
                counts[kind] = counts.get(kind, 0) + 1
            labels = [
                f"{count} {name}{'s' if count != 1 else ''}"
                for name, count in sorted(counts.items())
            ]
            parts.append("Shapes: " + ", ".join(labels) + ".")
        if text_regions:
            texts = [region.text for region in text_regions]
            parts.append("Text: " + "; ".join(texts) + ".")
        if latex:
            parts.append("Formulas: " + "; ".join(latex) + ".")
        return " ".join(parts)

File: packages/ai-assistant/src/providers/offline.ts
import { Provider, SummarizeOptions } from './base';

export interface DiagramShapeLike {
  kind?: string;
  label?: string;
}

export interface DiagramTextLike {
  text?: string;
}

export class RuleProvider implements Provider {
  name = 'offline';

  available(): boolean {
    return true;
  }

  async summarize(_systemPrompt: string, _userPrompt: string, _options?: SummarizeOptions): Promise<string> {
    return '';
  }

  static summarizeDiagram(
    textRegions: DiagramTextLike[],
    shapes: DiagramShapeLike[],
    latex: string[],
  ): string {
    if (shapes.length === 0 && textRegions.length === 0 && latex.length === 0) {
      return 'The canvas is blank: no shapes, text or formulas were recognized.';
    }

    const parts: string[] = [];
    if (shapes.length > 0) {
      const counts = new Map<string, number>();
      for (const shape of shapes) {
        const kind = shape.kind || 'unknown';
        counts.set(kind, (counts.get(kind) ?? 0) + 1);
      }
      const labels = [...counts.entries()]
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([name, count]) => `${count} ${name}${count !== 1 ? 's' : ''}`);
      parts.push(`Shapes: ${labels.join(', ')}.`);
    }
    if (textRegions.length > 0) {
      parts.push(`Text: ${textRegions.map((region) => region.text ?? '').join('; ')}.`);
    }
    if (latex.length > 0) {
      parts.push(`Formulas: ${latex.join('; ')}.`);
    }
    return parts.join(' ');
  }
}
